File: interop/go/ping.go
// NMP/1 Ping encoder; validate all CLI inputs before constructing JSON.
package main
import ("encoding/hex"; "fmt"; "os"; "regexp"; "strconv"; "strings")
func uint(s string) bool {
 if !regexp.MustCompile(`^(0|[1-9][0-9]{0,19})$`).MatchString(s) { return false }
 _, err := strconv.ParseUint(s,10,64); return err == nil
}
func main() {
 a:=os.Args[1:]
 if len(a)!=3 || !regexp.MustCompile(`^[0-9a-fA-F]{64}$`).MatchString(a[0]) || !uint(a[1]) || !uint(a[2]) || a[1]=="0" { fmt.Fprintln(os.Stderr,"expected: NODE_HEX MESSAGE_ID(1..u64) NONCE(0..u64)"); os.Exit(1) }
 bytes,_:=hex.DecodeString(a[0]); sender:=make([]string,32)
 for i,b:=range bytes {sender[i]=strconv.Itoa(int(b))}
 fmt.Printf("{\"version\":1,\"message_id\":%s,\"sender\":[%s],\"message\":{\"type\":\"Ping\",\"body\":{\"nonce\":%s}}}\n",a[1],strings.Join(sender,","),a[2])
}

File: interop/python/ping.py
"""Encode an NMP/1 Ping envelope; transport and authentication are separate."""
import json
import re
import sys

def uint(value):
    return re.fullmatch(r"0|[1-9][0-9]{0,19}", value) is not None and int(value) <= 18446744073709551615

args = sys.argv[1:]
if len(args) != 3 or re.fullmatch(r"[0-9a-fA-F]{64}", args[0]) is None or not all(map(uint, args[1:])) or args[1] == "0":
    sys.exit("expected: NODE_HEX MESSAGE_ID(1..u64) NONCE(0..u64)")
print(json.dumps({"version": 1, "message_id": int(args[1]), "sender": list(bytes.fromhex(args[0])), "message": {"type": "Ping", "body": {"nonce": int(args[2])}}}, separators=(",", ":")))
